src/mongocxx/database.hpp

```diff
@@ -248,6 +248,7 @@ class MONGOCXX_API database {
 
     class MONGOCXX_PRIVATE impl;
     std::unique_ptr<impl> _impl;
+    const std::unique_ptr<impl>& get_impl() const;
```

Comment:
The _impl field should definitely be a std::unique_ptr, I just don't think the getters should return that.  We're using std::unique_ptr here for lifetime management, so that _impl will get properly cleaned up when the owning database/collection/other object is destroyed.  Because of these semantics, a std::unique_ptr should only be held/used by one thing.  I'd implement the getters like so:
```
const impl& get_impl() const {
   // raise an error etc.
   return *_impl;
}
```
So instead, we return a reference/pointer to the underlying impl for temporary, non-owning access.  I kind of think of it like you're walking a dog, and only one person can hold the leash, but other people can stop and pet the dog.